src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -587,21 +591,17 @@ private MsalAccessTokenCacheItem FilterByKeyId(MsalAccessTokenCacheItem item, Au
                 .ConfigureAwait(false);
             var aliases = metadata.Aliases;
 
-            IEnumerable<MsalRefreshTokenCacheKey> candidateRtKeys = aliases.Select(
-                    al => new MsalRefreshTokenCacheKey(
-                        al,
-                        requestParams.AppConfig.ClientId,
-                        requestParams.Account?.HomeAccountId?.Identifier,
-                        familyId));
-
-            MsalRefreshTokenCacheItem candidateRt = allRts.FirstOrDefault(
-                rt => candidateRtKeys.Any(
-                    candidateKey => object.Equals(rt.GetKey(), candidateKey)));
+            allRts = FilterRtsByHomeAccountIdOrAssertion(requestParams, allRts, familyId);
+            allRts = allRts.Where(
+                item => aliases.ContainsOrdinalIgnoreCase(item.Environment));
 
-            requestParams.RequestContext.Logger.Info("Refresh token found in the cache? - " + (candidateRt != null));
+            IReadOnlyList<MsalRefreshTokenCacheItem> finalList = allRts.ToList();
+            requestParams.RequestContext.Logger.Info("Refresh token found in the cache? - " + (finalList.Count() != 0));
 
-            if (candidateRt != null)
-                return candidateRt;
+            if (finalList.Count() > 0)
```

Comment:
nit: `.Count` instead of `.Count()` (although since it's a list, in might not matter.